onnx/defs/rnn/defs.cc

```diff
@@ -5,124 +5,311 @@
+           "to be 0. It has shape `[batch_size, hidden_size]`.",
+	   true /*optional*/)	   
+    .Input(6, "initial_c",
+           "Optional initial value of the cell. If not specified - assumed "
+	   "to be 0. It has shape `[batch_size, hidden_size]`.",
+	   true /*optional*/)
+    .Input(7, "seq_lens",
+           "Optional tensor specifying lengths of the sequences in a batch. "
+           "If not specified - assumed all sequences in the batch to have "
+	   "length `seq_length`. It has shape `[batch_size]`.",
+	   true /*optional*/)	   
+    .Output(0, "output",
+	    "A tensor that concats all the intermediate output values of the "
+	    "hidden. It has shape `[seq_length, batch_size, hidden_size]`.")	    
+    .Output(1, "output_h",
+            "The last output value of the hidden. It has shape "
+	    "`[batch_size, hidden_size]`.");
+
+
+OPERATOR_SCHEMA(BidirectionalLSTM)
+    .NumInputs(3, 8)
+    .NumOutputs(1, 2)
+    .SetDoc(R"DOC(
+Computes an one-layer bidirectional LSTM. This operator is usually
+supported via some custom implementation such as CuDNN.
+
+Notations:
+`X` - input tensor
+`i` - input gate
+`o` - output gate
+`f` - forget gate
+`c` - cell gate
+`t` - time step (t-1 means previous time step)
+`W[iofc]` - W parameter weight matrix for input, output, forget, and cell gates
+`R[iofc]` - R recurrence weight matrix for input, output, forget, and cell gates
+`Wb[iofc]` - W bias vectors for input, output, forget, and cell gates
+`Rb[iofc]` - R bias vectors for input, output, forget, and cell gates
+`P[iof]`  - P peephole weight matrix for input, output, and forget gates
+`WR[iofc]` - WR parameter weight matrix for reverse direction input, output, forget, and cell gates
+`RR[iofc]` - RR recurrence weight matrix for reverse direction input, output, forget, and cell gates
+`WRb[iofc]` - WR bias vectors for reverse direction input, output, forget, and cell gates
+`RRb[iofc]` - RR bias vectors for reverse direction input, output, forget, and cell gates
+`PR[iof]`  - PR peephole weight matrix for reverse direction input, output, and forget gates
+`tanh(X)` - hyperbolic tangent of X
+`sigmoid(X)` - 1 / (1 + e^-X)
+`H` - Hidden state
+
+Equations (forward direction LSTM with default activations):
+  - it = sigmoid(Wi*Xt + Ri*Ht-1 + Wbi + Rbi)
+  - ft = sigmoid(Wf*Xt + Rf*Ht-1 + Wbf + Rbf)
+  - ot = sigmoid(Wo*Xt + Ro*Ht-1 + Wbo + Rbo)
+  - ct = tanh(Wc*Xt + Rc*Ht-1 + Wbc + Rbc)
+  - Ct = ft (.) Ct-1 + it (.) ct
+  - H = ot (.) tanh(Ct)
 )DOC")
-    .Input(1, "input",
+    .Attr("activations", "A list of 4 activation functions for input, output, "
+	  "forget, and cell gates. Typical activation functions are sigmoid "
+	  "and tanh. See the equations for default if not specified.",
+          AttrType::STRINGS)
+    .Attr("hidden_size", "Number of neurons in the hidden layer", AttrType::INT)
+    .Attr("input_forget", "Couple the input and forget gates if 1, default 0.",
```

Comment:
so if you couple them the input weight would be different? should we instead create a yet another op instead?

Reply:
My understanding is W_i is ignored in this case. I am inclined not to create a new op for it.

Reply:
I'd add this: the forget gate is computed as f = 1-i. The weights for the forget gate are ignored.